core/server/master/src/test/java/alluxio/master/file/FileSystemMasterTest.java

```diff
@@ -451,15 +453,142 @@ public void deleteDirRecursiveWithInsufficientPermissions() throws Exception {
+        .mergeFrom(SetAttributePOptions.newBuilder().setMode(new Mode((short) 0777).toProto())));
+    mFileSystemMaster.setAttribute(new AlluxioURI("/nested/test/dir/file"), SetAttributeContext
+        .mergeFrom(SetAttributePOptions.newBuilder().setMode(new Mode((short) 0777).toProto())));
+    // The user has permission over everything under this dir, so everything will be removed
+    mFileSystemMaster.setAttribute(new AlluxioURI("/nested/test2/file"), SetAttributeContext
+        .mergeFrom(SetAttributePOptions.newBuilder().setMode(new Mode((short) 0777).toProto())));
+    mFileSystemMaster.setAttribute(new AlluxioURI("/nested/test2/file2"), SetAttributeContext
+        .mergeFrom(SetAttributePOptions.newBuilder().setMode(new Mode((short) 0777).toProto())));
+    mFileSystemMaster.setAttribute(new AlluxioURI("/nested/test2/dir/file"), SetAttributeContext
+        .mergeFrom(SetAttributePOptions.newBuilder().setMode(new Mode((short) 0777).toProto())));
+
+    try (AuthenticatedClientUserResource userA = new AuthenticatedClientUserResource("userA",
+        ServerConfiguration.global())) {
+      mFileSystemMaster.delete(new AlluxioURI("/nested"),
+          DeleteContext.mergeFrom(DeletePOptions.newBuilder().setRecursive(true)));
+      fail("Deleting a directory w/ insufficient permission on child should fail");
+    } catch (FailedPreconditionException e) {
+      assertTrue(e.getMessage().contains("/nested/test (Permission denied"));
+      assertTrue(e.getMessage().contains("/nested (Directory not empty)"));
+    }
+    // The existing files/dirs will be: /, /nested/, /nested/test/ and everything under /nested/test/
+
+    assertNotEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(ROOT_URI));
+    assertNotEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/nested")));
+    assertNotEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(NESTED_URI));
+    assertNotEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(NESTED_FILE_URI));
+    assertNotEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/nested/test/file2")));
+    assertNotEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(NESTED_DIR_URI));
+    assertNotEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(NESTED_DIR_FILE_URI));
+    // The other files should be deleted successfully
+    assertEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/nested/test2")));
+    assertEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/nested/test2/file")));
+    assertEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/nested/test2/file2")));
+    assertEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/nested/test2/dir")));
+    assertEquals(IdUtils.INVALID_FILE_ID, mFileSystemMaster.getFileId(new AlluxioURI("/nested/test2/dir/file")));
```

Comment:
@tcrain See the 2 new unit tests for the behavior now. The behavior now is:
1. Check WRITE perm on all children files. I didn't change the perm model and left that task to https://github.com/Alluxio/alluxio/issues/15187 instead. Before this change, we check all WRITE perm, and after this change, we still check all WRITE perm on ALL children.
2. If any child fails the perm check, it will be left untouched. This propagates to the parent (and the parent's parent etc). The parent dir will then not deleted because it is not empty.

I actually fail to understand your example https://github.com/Alluxio/alluxio/pull/15113#issuecomment-1062081038 I don't really get what the difference is between a file directly under the target path, and a file in a nested dir. I mean I actually feel the "posix" way is more counter-intuitive. And that makes me a bit reluctant to fully conform. And my excuses of not doing this are:
1. I threw the POSIX compliance question to https://github.com/Alluxio/alluxio/issues/15187
2. HDFS itself is not fully POSIX compliant, so maybe we don't have to either?

@tcrain Regarding your example, I guess I don't understand why a nested dir will make things different. Could you shed some light on that? If it can be better explained (instead of just "POSIX said so") and if the logic is not too tricky, we can add it to this PR too.

Reply:
Discussed with @bobbai00 today and we think it's probably: When walking the FS tree sees a child, check its parent's permission and if not enough permission on the parent, throw an error. That explains why having the parent perm makes you able to delete files under `tmp/` but not having parent perm on `tmp/tmp2/` makes you unable to delete `tmp/tmp2/file`. 

Despite the discussions here, i'll try to follow the existing permission model as much as possible in this particular PR. If we think it's necessary, I'll create a separate PR to address this caveat https://github.com/Alluxio/alluxio/pull/15113#issuecomment-1062081038

WDYT @yuzhu @ggezer @madanadit ?

Reply:
Thanks!

Bob did a comparison of Alluxio vs HDFS vs Linux behavior for `rm -R` and perm requirement. Seems we are quite different. Moving this discussion to a separate dedicated issue:
https://github.com/Alluxio/alluxio/issues/15187#issuecomment-1078572463

So for this PR, the perm model will not be changed. Let's focus on the perf improvement without perm change. We take one step at a time and segregate the perm change in another PR.

Reply:
Agreed, let's not change the permission model in this PR.  any changes in the permission model needs a separate discussion.